amaranth/lib/io.py

```diff
@@ -253,6 +258,362 @@ def __repr__(self):
         return f"DifferentialPort({self._p!r}, {self._n!r}, invert={invert!r}, direction={self._direction})"
 
 
+class Buffer(wiring.Component):
+    """A combinatorial I/O buffer.
+
+    Parameters
+    ----------
+    direction : Direction
+    port : SingleEndedPort, DifferentialPort, or a custom platform-specific type
+
+    Attributes
+    ----------
+    signature : Buffer.Signature
+        Created based on constructor arguments.
+    """
+    class Signature(wiring.Signature):
+        """A signature of a combinatorial I/O buffer.
+
+        Parameters
+        ----------
+        direction : Direction
+        width : int
+
+        Attributes
+        ----------
+        i: :py:`unsigned(width)` (if :py:`direction in (Direction.Input, Direction.Bidir)`)
+        o: :py:`unsigned(width)` (if :py:`direction in (Direction.Output, Direction.Bidir)`)
+        oe: :py:`unsigned(1)` (if :py:`direction in (Direction.Output, Direction.Bidir)`)
+        """
+        def __init__(self, direction, width):
+            self._direction = Direction(direction)
+            self._width = operator.index(width)
+            members = {}
+            if self._direction is not Direction.Output:
+                members["i"] = wiring.Out(self._width)
+            if self._direction is not Direction.Input:
+                members["o"] = wiring.In(self._width)
+                members["oe"] = wiring.In(1, init=int(self._direction is Direction.Output))
+            super().__init__(members)
+
+        @property
+        def direction(self):
+            return self._direction
+
+        @property
+        def width(self):
+            return self._width
+
+        def __eq__(self, other):
+            return type(self) == type(other) and self.direction == other.direction and self.width == other.width
```

Comment:
I usually use `type(self) is type(other)` in case someone is doing weird shit with metaclasses.